File: h5gl/new/src/pan3d/res/SceneRes.ts
module Pan3d {
    export class SceneRes extends BaseRes {
        private _completeFun: Function;
        private _readDataFun: Function;
        protected _progressFun: Function;
        public sceneData: any;
        constructor(value: Scene3D) {
            super(value);
        }
        public load($url: string, $completeFun: Function, $progressFun: Function, $readDataFun: Function): void {
            this._completeFun = $completeFun;
            this._readDataFun = $readDataFun;
            this._progressFun = $progressFun;
            $url = this.scene3D.fileRoot + getMapUrl($url);
            LoadManager.getInstance().load($url, LoadManager.BYTE_TYPE, ($byte: ArrayBuffer) => {
                this.loadComplete($byte);

            }, null, $progressFun);
        }
        private loadComplete($byte: ArrayBuffer) {
            this._byte = new Pan3dByteArray($byte);

            this._byte.position = 0;
            this.version = this._byte.readInt()
            this.read(() => { this.readNext() });//img
        }
        private readNext(): void {
            this.read();//obj
            this.read();//material
            this.read();//particle;
            this.readScene();

            this._completeFun ();

        }
        private readScene() {
            var types: number = this._byte.readInt();
            this.readAstat();
            if (this.version >= 28) {
                this.readTerrainIdInfoBitmapData(this._byte)
            }
            var size: number = this._byte.readInt();
            this.sceneData = JSON.parse(this._byte.readUTFBytes(size));
 

        }
        private readAstat(): void {
            var hasAstat: boolean = this._byte.readBoolean();
            if (hasAstat) {


                this._byte.readFloat();
                this._byte.readFloat();
                this._byte.readFloat();
                this._byte.readFloat();
                var i: number;
                var j: number;
                var tw: number = this._byte.readInt();
                var th: number = this._byte.readInt();

                // this._astarDataMesh.width = tw;
                // this._astarDataMesh.height = th;
                if (this.version < 25) {
                    for (i = 0; i < th; i++) {
                        var tempAstar: Array<number> = new Array
                        for (j = 0; j < tw; j++) {
                            tempAstar.push(this._byte.readFloat())
                        }
                        // this._astarDataMesh.astarItem.push(tempAstar);
                    }
                    for (i = 0; i < th; i++) {
                        var tempHeightArr: Array<number> = new Array
                        for (j = 0; j < tw; j++) {
                            tempHeightArr.push(this._byte.readFloat())
                        }
                        // this._astarDataMesh.heightItem.push(tempHeightArr);
                    }
                } else {
                    var $heightScaleNum: number = this._byte.readFloat();
                    var $astrBase: Array<number> = this.readAstarFromByte(this._byte);
                    var $jumpBase: Array<number> = this.readAstarFromByte(this._byte);

                    var $astrBaseId: number = 0;
                    var $jumpBaseId: number = 0;
                    for (i = 0; i < th; i++) {
                        var tempAstar: Array<number> = new Array;
                        var tempJump: Array<number> = new Array;
                        for (j = 0; j < tw; j++) {
                            var astarNum: number = $astrBase[$astrBaseId++]
                            tempAstar.push(astarNum);
                            if (astarNum == 1) {
                                var ssss: number = $jumpBase[$jumpBaseId++]
                                tempJump.push(ssss);
                            } else {
                                tempJump.push(0);
                            }
                        }

                    }


                    for (i = 0; i < th; i++) {
                        var tempHeightArr: Array<number> = new Array;
                        for (j = 0; j < tw; j++) {
                            tempHeightArr.push(this._byte.readShort() / $heightScaleNum);
                        }

                    }

                }
            }
        }
        private readTerrainIdInfoBitmapData($byte: Pan3dByteArray): void {
            var $len: number = $byte.readInt();
            if ($len) {
                //var newByte: ByteArray = new ByteArray();
                //newByte.length = $len;
                //$byte.readBytes(newByte, 0, $len);

                var zipLen: number = $len
                var aryBuf: ArrayBuffer = $byte.buffer.slice($byte.position, $byte.position + zipLen);
                $byte.position += zipLen;
                var zipedBuf: ArrayBuffer = unZip(aryBuf)
                var newByte: Pan3dByteArray = new Pan3dByteArray(zipedBuf);


            }

        }
        private readAstarFromByte($byte: Pan3dByteArray): Array<number> {
            var $len: number = $byte.readUnsignedInt();
            var $intLen: number = Math.ceil($len / 32);
            var $astrBase: Array<number> = new Array
            for (var i: number = 0; i < $intLen; i++) {
                var $num: number = $byte.readUnsignedInt();
                for (var j: number = 0; j < 32; j++) {
                    var $ast: number = $num & 1;
                    if ($astrBase.length < $len) {
                        $astrBase.push($ast);
                    }
                    $num >>= 1;
                }
            }
            return $astrBase
        }

    }

}

File: android/app/src/main/java/com/pan3d/res/SceneRes.java
package com.pan3d.res;

import android.util.Log;

import com.pan3d.base.ByteArray;
import com.pan3d.base.CallBackFun;
import com.pan3d.base.Scene_data;
import com.pan3d.scene.Scene3D;
import com.pan3d.units.LoadBackFun;
import com.pan3d.units.LoadManager;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;


public class SceneRes extends BaseRes {

    public  JSONObject  sceneData;
    private CallBackFun sceneFinishFun;
    public  SceneRes(Scene3D val  ){
        super(val);

    }
    public void load(String url ,final CallBackFun bfun) {
        scene3D.loadManager.loadUrl(Scene_data.fileRoot+url, LoadManager.BYTE_TYPE, new LoadBackFun() {
            @Override
            public void bfun(HashMap dic) {
                if(dic!=null){
                    ByteArray temp=(ByteArray)dic.get("byte");
                    loadComplete(temp,bfun);
                }
            }
        },null);
    }
    public void  loadComplete(ByteArray buffByte, CallBackFun bfun)
    {
        this.sceneFinishFun=bfun;
        this._byte =buffByte;
        this.applyByteArray();
    }
    public void applyByteArray() {

        this.version = this._byte.readInt();
        Log.d("SceneRes版本 ->",   this.version+"===> " );
        this.read(new CallBackFun() {
            @Override
            public void StateChange(boolean State) {
                readNext();
            }
        });
    }
    private void   readNext()
    {
        read();
        read();
        read();

        this.readScene();


    }
    private void  readScene()
    {
        int types = this._byte.readInt();
        this.readAstat();

        if (this.version >= 28) {
            this.readTerrainIdInfoBitmapData(this._byte);
        }
        int size   = this._byte.readInt();
        try {
            this.sceneData= new JSONObject(this._byte.readUTFBytes(size));

        } catch (JSONException e) {
            e.printStackTrace();
        }
        Log.d("partic地址 ->",   "aa");


        this.sceneFinishFun.StateChange(true);

    }
    private void readTerrainIdInfoBitmapData(ByteArray $byte) {
        int $len = $byte.readInt();
        if ($len>0) {
            byte[] objByte= $byte.readBytes($len);

        }

    }

    private void readAstat() {
        boolean hasAstat = this._byte.readBoolean();
        if (hasAstat) {

            float midu = this._byte.readFloat();
            float aPosx = this._byte.readFloat();
            float aPosy = this._byte.readFloat();
            float aPosz = this._byte.readFloat();
            int i ;
            int j;
            int tw = this._byte.readInt();
            int th = this._byte.readInt();

            if (this.version < 25) {
                for (i = 0; i < th; i++) {

                    for (j = 0; j < tw; j++) {
                        this._byte.readFloat();
                    }

                }
                for (i = 0; i < th; i++) {

                    for (j = 0; j < tw; j++) {
                        this._byte.readFloat();
                    }

                }
            } else {
                this._byte.readFloat();
                this.readAstarFromByte(this._byte);
                this.readAstarFromByte(this._byte);



                for (i = 0; i < th; i++) {

                    for (j = 0; j < tw; j++) {
                        this._byte.readShort()  ;
                    }

                }

            }
        }
    }
    private void readAstarFromByte(ByteArray $byte){
        int $len = $byte.readUnsignedInt();
        int $intLen =(int) Math.ceil($len / 32.0f);
        for (int i = 0; i < $intLen; i++) {
            $byte.readUnsignedInt();

        }

    }



}
